main/tests/server/src/com/google/refine/exporters/TsvExporterTests.java

```diff
@@ -171,7 +171,7 @@ public void exportTsvWithQuote() {
 
         Assert.assertEquals(writer.toString(), "column0\tcolumn1\tcolumn2\n" +
                 "row0cell0\trow0cell1\trow0cell2\n" +
-                "row1cell0\t\"line has \"\"quote\"\"\"\trow1cell2\n" +
+                "row1cell0\t\"line has \"quote\"\"\trow1cell2\n" +
```

Comment:
This new test case seems incorrect to me. If we do not have two double quotes inside the cell value, then the cell value should not be quoted at all. So it should be:
```suggestion
                "row1cell0\tline has \"quote\"\trow1cell2\n" +
```

But I suspect it will be hard to get this behaviour without modifying CSVWriter itself.

Reply:
Two double quotes are removed inside the cell value to make sure that double quotes are not escaped. However, the  quote outside cell value are added in the assert statement as the "cells.set" method is preserving the double quotes as depicted in screenshot attached. 


I would also like to add that I have manually also tested these cases in openrefine locally after making the changes and it is working in all these scenarios.
![TSV test](https://user-images.githubusercontent.com/59418361/164581021-f63598db-4dbb-4c5d-bcc6-91758e26a267.png)

Reply:
Sorry I still think the output specified by this test case is wrong (see the original issue).

Here is another test case worth thinking about: exporting a cell which contains a quote character followed by a TAB character. In this case, we do need the quotes around the cell value, and therefore we also need escaping the quote inside the cell value, right?

Reply:
I have tried this scenario also. The input text copied on clipboard with various combination in cell such as `,",\t is attached. The local tsv export of that project is also enclosed (.txt is added at the end just to attach it here).
![input](https://user-images.githubusercontent.com/59418361/164736153-1abbd74e-27f0-4244-89f9-099fc28df1b2.png)

[tsv_exportfile.tsv.txt](https://github.com/OpenRefine/OpenRefine/files/8541809/tsv_exportfile.tsv.txt)

If you still feel this is not right, I will try implementing CSV Writer but I strongly believe it meet the requirements as mentioned in the issue
Thanks & Warm Regards

Reply:
I think what is really important to keep in mind is how a program can parse the exported file back into a table. For instance, OpenRefine itself should normally be able to parse a file it exported itself back as a table (and ideally other programs, such as LibreOffice or Excel).

If I open your exported TSV file above in OpenRefine, I cannot find a set of options which would give me the original table. If I disable escaping, I get this:
![image](https://user-images.githubusercontent.com/309908/164738399-034b4219-3e2a-457a-8e20-e0654f93fe20.png)

Did I miss anything here?

Reply:
Thanks for looking at it. Here, the spitted column has a tab character which i have manually inserted just to check various combination which is spitted into two columns by the **CSVImporter.** Tab character here is the special case case while importing however  there is no other issue in any other character.

Warm Regards

Reply:
The exported file should be parsable regardless of the characters contained in the cells.

Reply:
@TheTrailblazer09: I think this is a situation where you gain a lot from writing your test case before coding. Try to find a few bizarre cases. I would be interested to see if a cell that is already in double quotes, would have them escaped before having normals double quotes being added.

Regards, Antoine

Reply:
Hi,
I have added a test case where the cell is enclosed by the quote character and the CSVWriter seems to pass the test. Below is the output produced in both scenarios for ready reference:-

Input:  "cell enclosed by (") quote character"

Output produced without escaping:    """cell enclosed by ("") quote character"""

Output produced with escaping:  ""cell enclosed by (") quote character""

Regards,
Vani

Reply:
> Input: "cell enclosed by (") quote character"
> 
> Regards, Vani

Yes, that's one scenario @TheTrailblazer09. I'm talking about a lots of scenarii…

1. "cell enclosed by (") double quote character"
2. cell containing (") double quote character
3. "cell starting with a double quote character
4. cell ending with a double quote character"
5. cell containing (") double quote and (`\t`) tab character
6. `\t`cell containing (") double quote and starting with (`\t`) tab character
7.  `\t`cell enclosed by (`\t`) tab character`\t`
8. …
9. cell containing (') single quote character
10. …
11. cell containing (`) single quote character
12. 

It's really to save you time. There are so many time this is used, so it used with so much different data, that more test seams to me a saving net. @wetneb, how would you feel about that?

Regards, Antoine

Reply:
@antoine2711  Added 25 additional test cases. Kindly review them.

Thanks & Warm Regards

> > Input: "cell enclosed by (") quote character"
> > Regards, Vani
> 
> Yes, that's one scenario @TheTrailblazer09. I'm talking about a lots of scenarii…
> 
> 1. "cell enclosed by (") double quote character"
> 2. cell containing (") double quote character
> 3. "cell starting with a double quote character
> 4. cell ending with a double quote character"
> 5. cell containing (") double quote and (`\t`) tab character
> 6. `\t`cell containing (") double quote and starting with (`\t`) tab character
> 7. `\t`cell enclosed by (`\t`) tab character`\t`
> 8. …
> 9. cell containing (') single quote character
> 10. …
> 11. cell containing (`) single quote character
> 
> It's really to save you time. There are so many time this function is used by the users and in so mamny contexes, and it used with so much different data, that more test seams to me like a saving net. @wetneb, how would you feel about that?
> 
> Regards, Antoine

@antoine2711  Added 25 additional test cases. Kindly review them.

Thanks & Warm Regards
Vani

Reply:
@wetneb: just to make sure... Here, did I suggest well in asking more tests, and is this too much, or correct? Maybe the caret tests are too much...

Also, is it possible/better to regroup them instead of having a single test for each 20+?

Regards, Antoine

Reply:
@antoine2711  @wetneb  tests are vital .

> did I suggest well in asking more tests, and is this too much, or correct?

Reply:
> @wetneb: just to make sure... Here, did I suggest well in asking more tests, and is this too much, or correct? Maybe the caret tests are too much...
> 
> Also, is it possible/better to regroup them instead of having a single test for each 20+?
> 
> Regards, Antoine

Hi
@Antoine. In CI/CD some tests have failed which shows some I/O error. However, all these tests have passed locally (./refine test). could you please guide me on this if anything is needed from my side.

Thanks & Warm Regards
Vani